### PR TITLE
Alerting: Do not persist noop transition from Normal state.

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What is this feature?**
This PR updates the alerting state manager 
1. to skip saving state transitions `Normal (no reason) -> Normal (no reason)`. This feature is controlled by feature flag `alertingNoNormalState`
2. to not return `Normal (no-reason)` state to GET API, which makes UI show only states that are not Normal. Rules that have normal state are still shown.
<details> <summary> Screenshot </summary>

image

</details> 

**Why do we need this feature?**
Every time when each rule is evaluated state manager calculates a new state. In the case of multi-dimensional rules, it creates many states, depending on how many series are returned by the evaluation engine. Sometimes it can be 1000s of states per rule.
Also, it updates the database and saves (upserts) all states touched at this evaluation cycle (usually, it's all states that belong to a rule). It saves even states that were not changed, e.g. transition from state `Alerting -> Alerting` or `Normal (no-reason) -> (no-reason)`. The only purpose of saving states to the database is recovery after restart (or in the case of HA mode, synchronizing the state of a new Grafana instance with ones that are already running). However, for example, the Normal state is not used for anything but the presentation - the UI displays them.

This operation together with reading data source info is by far the most consuming of the database connection pool. When there are thousands of alert rules evaluated, it makes the system starving for connections. It affects the regular user queries, and alerting scheduler because it can start lagging behind, dropped evaluations, etc.

According to the statistics I gathered in the Grafana cloud, more than 90% of alert states are Normal. 

Therefore, if Grafana stops sending noop transitions (i.e. Normal -> Normal) to the database, it will greatly reduce the contention in the connection pool, as well as reduce traffic to the database.

**Who is this feature for?**
This feature will be especially useful for Grafana instances that handle many thousands of alert rules 

### Benchmark results ###
Test environment:
- Kubernetes. 
- Grafana container on a node with up to 8 CPU and 32Gb memory (no limits, no other containers). Enabled tracing.
   Settings `max_idle_conn` and `max_open_conn` set to `20`.
- MySql container (another node) of the same configuration.
- 25000 alert rules with an evaluation interval of 30s. The data source is `Test Datasource` with RandomWalk + threshold that makes sure that state is *always* Normal.

Key takeaways:
- The average rule evaluation duration was reduced from 2.5s to 500ms (99th percentile from 18s to 5s)
- Time spent on waiting for SQL connection dropped more than 5 times.
- No significant scheduler lag (when the tick timestamp is different than real clock)
- Traffic to the database dropped 2x times
- 15% less memory
- 25% less CPU usage.

<details><summary> Graphs  </summary>

Alerting engine (PR, no normal state):
image

Alerting engine (all states)
image

Alerting engine (composite)
image

Grafana SQL client stats (PR, no normal state):
image**

Grafana SQL client stats (all states)
image

Grafana SQL client stats (composite)
image

Container Grafana (composite)
image

Networking Mysql (composite)
image

</details>

### CODE REVIEW

File: docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md [hunk 1]
```diff
@@ -45,6 +45,7 @@ Some stable features are enabled by default. You can disable a stable feature by
 | `autoMigrateGraphPanels`          | Replace the angular graph panel with timeseries                                 |
 | `datasourceLogger`                | Logs all datasource requests                                                    |
 | `accessControlOnCall`             | Access control primitives for OnCall                                            |
+| `alertingNoNormalState`           | Stop maintaining state of alerts that are not firing                            |
```
Comment (gotjosh) [inline docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md:48]: Can we artificially assume this state instead of putting this behind a feature flag then? As in, could we assume (and insert at runtime) a new state that would assume normal when a previous state its state is missing.

Comment (yuri-tceretian) [reply docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md:48]: we can artificially assume a state for a rule, but we cannot assume a state for each rule's instances when Grafana starts. The normal state will be available only after the first evaluation once Grafana starts, which can be in 1s or 1h or more.

Comment (yuri-tceretian) [reply docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md:48]: I mean this from a presentation point of view. From state management pov we already do this (i.e. artificially assume a new instance to be in normal state. The difference is in presentation: currently, we show instances with normal state in UI, with this feature flag enabled, we are not able to show until the first evaluation after the restart. Therefore, I decided to not show it at all because it could be confusing (especially in environments when Grafana can be restarted at any time, like K8s)

File: pkg/services/ngalert/schedule/schedule_unit_test.go [hunk 2]
```diff
@@ -691,7 +691,7 @@ func setupScheduler(t *testing.T, rs *fakeRulesStore, is *state.FakeInstanceStor
 		Tracer:           testTracer,
 	}
 
-	st := state.NewManager(m.GetStateMetrics(), nil, is, &state.NoopImageService{}, mockedClock, &state.FakeHistorian{})
+	st := state.NewManager(m.GetStateMetrics(), nil, is, &state.NoopImageService{}, mockedClock, &state.FakeHistorian{}, false)
```
Comment (alexweav) [inline pkg/services/ngalert/schedule/schedule_unit_test.go:694]: Can we refactor the constructor args to a struct like `state.ManagerOptions`, kind of like we did for the scheduler? https://github.com/grafana/grafana/blob/2505f112f5e140e7ba45be82e27ecfc977a5aeea/pkg/services/ngalert/schedule/schedule.go#L104

The boolean argument that significantly changes the behavior is a bit confusing IMO.

File: pkg/services/ngalert/state/cache.go [hunk 3]
```diff
@@ -209,19 +209,22 @@ func (c *cache) get(orgID int64, alertRuleUID, stateId string) *State {
 	return nil
 }
 
-func (c *cache) getAll(orgID int64) []*State {
+func (c *cache) getAll(orgID int64, skipNormalState bool) []*State {
```
Comment (alexweav) [inline pkg/services/ngalert/state/cache.go:212]: I wonder if we should dependency inject all of `FeatureToggles` rather than passing this bool around everywhere. I realize it's more overhead, but the bool hurts readability at call sites IMO.

And, if we decide we like this change and want to make it always-on (removing the toggle), it's going to be a pain to dig back through all this code and remove the args.

File: pkg/services/ngalert/state/manager.go [hunk 4]
```diff
@@ -37,19 +37,22 @@ type Manager struct {
 	images        ImageCapturer
 	historian     Historian
 	externalURL   *url.URL
+
+	doNotKeepNormalState bool
 }
 
-func NewManager(metrics *metrics.State, externalURL *url.URL, instanceStore InstanceStore, images ImageCapturer, clock clock.Clock, historian Historian) *Manager {
+func NewManager(metrics *metrics.State, externalURL *url.URL, instanceStore InstanceStore, images ImageCapturer, clock clock.Clock, historian Historian, doNotKeepNormalState bool) *Manager {
```
Comment (alexweav) [inline pkg/services/ngalert/state/manager.go:44]: `doNotKeepNormalState` == `skipNormalState`? This is another reason why i'd prefer just using the toggle directly, it's less ambiguous and less intrusive to the existing code.

File: pkg/services/ngalert/state/cache.go [hunk 5]
```diff
@@ -209,19 +209,22 @@ func (c *cache) get(orgID int64, alertRuleUID, stateId string) *State {
...
+			}
 			states = append(states, v2)
 		}
 	}
 	return states
 }
 
-func (c *cache) getStatesForRuleUID(orgID int64, alertRuleUID string) []*State {
+func (c *cache) getStatesForRuleUID(orgID int64, alertRuleUID string, skipNormalState bool) []*State {
 	var result []*State
```
Comment (alexweav) [inline pkg/services/ngalert/state/cache.go:228]: Can we kill this line and just `result := make([]*State, 0, len(rs.states))` on 239?

File: pkg/services/ngalert/state/cache.go (hunk 3 above)
Comment (gotjosh) [reply pkg/services/ngalert/state/cache.go:212]: Per my comment on https://github.com/grafana/grafana/pull/61201#discussion_r1066034793 I was hoping we could make an argument for removing the feature flag w/o necessarily incurring in a breaking change. 

I think the feature flag makes this more complicated and doesn't really buy us anything (it is a breaking change but a very subtle one) - could we maybe assume that until determined by the next evaluation when a rule starts it's normal in order to keep semantics equal? The way I understand this if an instance would have no state to recover from it's because it was either:

A) never present 
B) not recorded because it was normal

Comment (alexweav) [reply pkg/services/ngalert/state/cache.go:212]: Sure, that would solve my comment here too, I'm just not a fan of the boolean flags permeating the code that deeply change its behavior. If we do this a couple times, we end up with a large matrix of boolean toggles and things become hard to follow. Just having one path is a solution here - but if we must have two, I'd rather it be flagged each place it's used.

Comment (yuri-tceretian) [reply pkg/services/ngalert/state/cache.go:212]: My intention is to keep the feature flag until Grafana 10 when we can reason about breaking changes. Although there is only one breaking change, in the presentation, I think it is still a breaking one and should be handled accordingly - we should make it default in the next major version. 

re features toggle, I am really on the fence. I did that initially but then decided to use the flag because backtesting creates a state manager and therefore it needs a feature manager just to read that toggle, and create a fake one was kinda messy, as well as propagating feature manager. Anyway, I see your point, and will update to use toggles.

File: pkg/services/ngalert/schedule/schedule_unit_test.go (hunk 2 above)
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule_unit_test.go:694]: Oh, I actually like this approach. Will do it

File: pkg/services/ngalert/state/manager.go (hunk 4 above)
Comment (yuri-tceretian) [reply pkg/services/ngalert/state/manager.go]: renamed to DoNotSaveNormalState, which I think more close to what it does

File: pkg/services/ngalert/state/cache.go (hunk 3 above)
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/state/cache.go:212]: I strongly agree with Alex's remark about using a boolean parameter to change the behavior of a function. That's a low-hanging fruit and we should change it. Have two functions or use the options pattern.

Comment (yuri-tceretian) [reply pkg/services/ngalert/state/cache.go:212]: If we abstract from the fact that this is a feature flag, the argument here makes perfect sense: a caller requests all abnormal states. I can agree that it could look obscure and unobvious in a public API like the constructor method of the state manager. However, I do not see any problems with methods, especially when it is the only boolean argument.
Providing an entire feature manager is what is very ambiguous and obscure because it makes it unclear what flags and how they are used. It is like providing an entire Grafana config just to read one field two layers down.

Comment (alexweav) [reply pkg/services/ngalert/state/cache.go:212]: > Providing an entire feature manager is what is very ambiguous

I'd agree with this for a lot of objects, but generally a feature flag structure intentionally reverses this rule:
1. It's clear at each point the flag is checked that it's a flagged code, not a permanent config toggle.
2. It makes the flag checks less intrusive, and easier to remove in the future.

Especially passing the boolean as part of the function signature, this will have to go through back and be refactored out later when we decide to drop the toggle and make this default behavior.

File: pkg/services/ngalert/state/state.go [hunk 6]
```diff
@@ -140,6 +140,11 @@ func (a *State) Maintain(interval int64, evaluatedAt time.Time) {
 	a.EndsAt = nextEndsTime(interval, evaluatedAt)
 }
 
+// IsNormalState returns true if the state is Normal and reason is empty
+func IsNormalState(s *State) bool {
```
[the remark this answers is not in the document]
Comment (yuri-tceretian) [reply pkg/services/ngalert/state/state.go]: Makes sense. Updated!

File: pkg/services/ngalert/state/cache.go (hunk 3 above)
Comment (yuri-tceretian) [reply pkg/services/ngalert/state/cache.go:212]: I still think that the bool argument is better than the feature manager in this particular case because the cache implementation has different methods like `getOrCreate`, `removeByRuleUID`, `get` that return state(s). Conceptually, having a feature flag changing the behavior of just a subset of methods would make it less clear what is going on and why different methods of the cache return different data. 
the flag there is purely for optimization purposes - to avoid filtering the slice in the manager.

**Review by brendamuir — APPROVED** [review]

**Review by alexweav — APPROVED** [review]
